staging_docs/user/guides/06-prune.md

```diff
@@ -0,0 +1,139 @@
+# Prune Repository Content
+
+A workflow that can be useful for specific kinds of installation is the "prune" workflow.
+For repositories that see frequent updates followed by long periods of stability, it can
+be desirable to eventually "age out" RPMs that have been superceded, after a period of time.
+
+The `/pulp/api/v3/rpm/prune/` API exists to provide to the repository-owner/admin a tool to
+accomplish this workflow.
+
+The `repo_hrefs` argument allows the user to specify a list of specific `RpmRepository` HREFs, or 
+the wildcard "*" to prune all repositories available in the user's domain.
+
+The `keep_days` argument allows the user to specify the number of days to allow "old" content to remain in the
+repository. The default is 14 days.
+
+The `repo_concurrency` argument allows the user to control how many `pulpcore-workers` can be operating on
```

Comment:
There are already a ton of ways for a single user to stuff up the dispatch process. I can, right now, create 1000 repos all pointing to the same remote, and then sync all of them in an endless loop.

I considered (and even at one point implemented) a similar percentage-setting. The prob w/ a setting is that it requires restarting all the workers to pick it up if you decide that you want something different, and can't be changed for just a specific run.

I *also* toyed with making this an admin-only command. That would answer the immediate COPR need. But our other future-looking service-needs "feel" like it needs to be accessible to end-users.

I think the "a single user can consume All The Workers" is a very valid issue - but it needs to be fixed at a more-foundational level than in individual calls. I'd rather see that addressed as its own feature for the entire system.

Reply:
I agree with your decision making the feature available to end-users.

However, this is not mutually exclusive - i am still leaning towards the % worker setting as we did for the import/export https://github.com/pulp/pulpcore/blob/072e43ffffc36e6cf05f13450f64aaeab7e5350b/pulpcore/app/settings.py#L328. Maybe we can re-purpose it and instead of making it purely import task specific make it generic enough so all tasks (import, prune,etc) would not use more that % of workers.
I understand that there are million ways how a user can just 'ddos' pulp system and the example with creating 1000 repos and sync is valid. But it is different from prune operation. Syncs  are separate 1000 tasks that will be created manually( in a sense that it will hit consciously 1000 times api calls)  one by one by the user. This situation is different because this is a *one* task (group) that will automatically in a glimpse of a moment will dispatch potentially as many or more tasks than workers.
I am ok with this being an admin setting, and if admin decides that group tasks somehow occupy more resources than desired than admin could just make that call and change it. WDYT?
I am not going to stand too much in the way, but it makes me feel slightly uncomfortable that we are going to allow a user to set the concurrency for the task group.

Reply:
Limiting the concurrency based on task priority, perhaps?

Reply:
To add to this list...replication feature probably should also be 'caped' because it also leverages the task group. I understand that this is probably separate discussion, but i somehow get more and more convinced that any feature that uses a task group should probably have the option to be 'caped' by admin.

Reply:
talking about 'priority' is a different level of complexity. Limiting-concurrency still maintains FIFO - 'priority' implies letting some task "go to the front of the line", and that combined with "but what if the new prio-1 task wants the same resource as an older prio-whenever task" makes it..fraught. And prone to deadlocks :)

Reply:
I'm not suggesting a generic priority system.  Just maybe "background level tasks only get to take up xx% of the workers at a time"

Reply:
> I'm not suggesting a generic priority system. Just maybe "background level tasks only get to take up xx% of the workers at a time"

Ah ok, cool, we're on the same page then

We can start with applying this to just task-group-issued tasks, I think. As Ina pointed out, if you're just dispatching tasks one at a time via API calls, things can interleave. The issue w/ task-groups doing this, is all the tasks the group dispatches get put in the queue *all at once*, exacerbating the problem.

Reply:
Investigation shows that "percentage of Workers.online()" can only be answered "inside" pulpcore - Worker isn't exposed to plugins (probably correctly). I'm in the process of:
- removing repo_concurrency from the API call (long-term, we don't want it there but rather in some more-consistent place like task-group-attribute/setting)
- Adding a `PRUNE_WORKERS = 5` setting to pulp_rpm
- teaching tasks/prune to look for `PRUNE_WORKERS`, and default to 5 if it doesn't exist.

That will accomplish:
- not letting Random User use up All The Workers At Once,
- letting the instance-admin decide how many workers they're willing to dedicate to prune,
- not requiring an API change when we provide a better answer for this at the pulpcore level - we'll just start ignoring the setting when the Better Way comes along, and
- allow us to provide this functionality in a useful way, now, instead of gating on design/implementation work from pulpcore.

Thoughts anyone?

Reply:
I was thinking we would add a "throttleable" or "background" flag to `dispatch()` and core would handle the worker percentages

Reply:
There are probably a half-dozen ways to approach it, and at least two or three will be good ones :)  But yeah, not for this PR.

Reply:
@ggainey I am down with this plan. I will bring up the topic of limiting workers usage for the task-group-issued tasks